### PR TITLE
[Gekidou] MM-39736 Login Flow

### PR DESCRIPTION
**Pull request by jfrerich**
#### Remaining work
- [ ] Test previous components that import <Loading> component and verify expected outcome
    The previous instances all used the default values for the size and color.  (see diff of <Loading> component file)

### Summary
-  Refactor SSO login options
    -  remove `gitlab.tsx`, `google.tsx`, `office365.tsx`, `saml.tsx` and `open_id.tsx`
    -  create `sso_options.tsx` with logic for creating SSO buttons 
       - depends on # of buttons available
-  `Background.tsx` out of server folder
- add new icons for SSO options
- Loading component now uses a Lottie file. 
  - update ios and android folders to link `lottie-react-native`
  - move from `channel_list` component and replace in `component/loading`
- `login_options_separator.tsx` text separator with horizontal lines
  - this could be shared elsewhere
- `buttonStyles.ts` 
	- `lg` size was cutting off letters with lower tails (ie. `g`)
	- `textAlignVertical` is not a valid argument to `<button>`

**Non related PR changes**
-  don't autocapitalize the first letter or server URL and display name
- `buttonStyles.ts`
  - Before: 
    - 
  - After:
    - 

#### Different option combinations
- [ ] only email / no SSO (only email buttons shown
- [ ] only SSO / no email (only SSO buttons shown / `Continue with` text prepended)
- [ ] no email or SSO (no options screen)
- [ ] only 1 SSO (redirect to SSO URL and bypass login options screen
- [ ] 2 SSOs with email option (two SSO buttons are side by side) 
- [ ] > 2 SSO (all buttons stacked vertically)

#### Scrolling
- [ ] many SSO and email options (need to be able to scroll)

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note

```

### CODE REVIEW

File: app/screens/login/index.tsx
Comment (jfrerich) [inline app/screens/login/index.tsx:299]: This shoudl not be white and should use lottie

File: app/screens/login/index.tsx
Comment (jfrerich) [inline app/screens/login/index.tsx:262]: What was this for?

File: app/screens/login/index.tsx
Comment (jfrerich) [inline app/screens/login/index.tsx:383]: address this

File: app/screens/login_options/sso_options.tsx
Comment (jfrerich) [inline app/screens/login_options/sso_options.tsx:1]: This file allows adding SSO options and removes all of the `<SSO_Option>.tsx` files.

File: app/constants/sso.ts
Comment (enahum) [inline app/constants/sso.ts:42]: this should no be included in this constants file.. why? cause the constants are initialized and we require the images even if the user won't ever use SSO or even after the user has logged in, thus loading assets and kept in memory without need.

**Review by enahum — APPROVED** [review]

**Comment by matthewbirtch** [comment]
Adding @Rina-dsg to do the first pass UX Review. Tag me after that if you want my review after that

**Comment by Rina-dsg** [comment]
1. it seems that the body text size a bit smaller than in design
2. status bar isn't visible (white)
Screenshot 2021-11-12 at 20 45 46

3. the vertical alignment of the content block should be centred. 
Screenshot 2021-11-12 at 21 02 35

**Comment by enahum** [comment]
@Rina-dsg the text perhaps.

The veryical alignment I'm afraid you're going to break functionality as it needs to allow scrolling up when the text inouts are in focus, to align it, it would need to take the entire vertical space thus preventing the scroll or it will overscroll badly, I did face those two situations when centering the content, but we could give it another try and see if we can find an alternate solution no promises.

That said, I'm responding to this even though you are talking about the server screen and not really anyhing related to the work done in this PR :) but I guess the subsequent screens will have the same feedback. More importantly it does not look centered on Figma either.

**Comment by Rina-dsg** [comment]
@enahum got it. Let's leave it as it is (without adjusting the vertical alignment).

1. arrow back and status bar are barely visible
2. the horizontal padding (right & left) should be 24pt
3. `Forgot password` shouldn't have opacity
4. `or log in with` seems like smaller than in design & `divider` should have `16% centre channel` (currently it seems a bit darker)
5. 'SAML' -> colors are different from the design

Screenshot 2021-11-12 at 20 43 41

**Comment by enahum** [comment]
> or log in with seems like smaller than in design 

@Rina-dsg yes, it seems that Jason tried to follow the Body 25 instead of using the actual values, that string though is using a size not defined in the design system, as a few others.. I'll make the change as in the design

**Comment by enahum** [comment]
@Rina-dsg 

image


theme sidebar bg does not really work well here with dark theme, so I'm going to change it to use the same color as the text

**Comment by enahum** [comment]
UX feedback addressed

File: app/screens/login/form.tsx
Comment (avinashlng1080) [inline app/screens/login/form.tsx:178]: If `loginError` is of type `Error`, then we won't have the `server-error_id` property.  
I think that this function can be flattened but 0/5 .
```
        const errorId = serverError.server_error_id;
```

**Review by avinashlng1080 — CHANGES_REQUESTED** [review]
@jfrerich  I have left a few non-blocking comments.

File: app/constants/sso.ts
Comment (enahum) [reply app/constants/sso.ts:10]: any strong reason why it needs to be sorted?

File: app/components/channel_list/index.tsx
Comment (enahum) [reply app/components/channel_list/index.tsx:18]: everywhere.. but in this case the comment is just to simplify showing/hiding some functionality in another PR, the fact is it was done as a component for that part of the code only and this PR made it so it is reusable.

File: app/screens/login/login_options_separator.tsx
Comment (enahum) [reply app/screens/login/login_options_separator.tsx:51]: 0/5

File: app/screens/login/sso_options.tsx
Comment (enahum) [reply app/screens/login/sso_options.tsx:81]: sso is never underfined

File: app/screens/login/index.tsx
Comment (enahum) [reply app/screens/login/index.tsx:85]: not really, the `t` helper function is used only in places where the ast won't be able to detect that the string is localizable.

File: app/screens/login/index.tsx
Comment (enahum) [reply app/screens/login/index.tsx:94]: same as the other answer

File: app/screens/login/form.tsx
Comment (enahum) [reply app/screens/login/form.tsx:94]: Android uses `windowMode=resize` which already takes care of moving the elements on screen when the keyboard is opened.

File: app/screens/login/form.tsx
Comment (enahum) [reply app/screens/login/form.tsx:178]: what would be your suggestion? your worry is being addressed on line 180

File: app/components/channel_list/index.tsx
Comment (avinashlng1080) [reply app/components/channel_list/index.tsx:18]: Alright, thanks.

File: app/screens/login/form.tsx
Comment (avinashlng1080) [reply app/screens/login/form.tsx:94]: Understood thanks :)

File: app/constants/sso.ts
Comment (avinashlng1080) [reply app/constants/sso.ts:10]: I think it's my bad, I developed this habit of sorting everything and also saw it on some occasions in our code.  But this is not a code standard and we can dismiss this comment.

**Review by koox00 — APPROVED** [review]

**Comment by matthewbirtch** [comment]
@Rina-dsg can you re-review? It looks like your feedback was addressed

**Review by Rina-dsg — APPROVED** [review]

File: app/utils/theme/index.ts
Comment (jfrerich) [reply app/utils/theme/index.ts]: Done.

File: app/screens/login/login_options_separator.tsx
Comment (jfrerich) [reply app/screens/login/login_options_separator.tsx]: done.

File: app/screens/login/sso_options.tsx
Comment (jfrerich) [reply app/screens/login/sso_options.tsx]: done.

File: app/screens/server/index.tsx
Comment (jfrerich) [reply app/screens/server/index.tsx]: This has been addressed. Note the code now exists in `apps/screens/server/index.tsx`

**Comment by jfrerich** [comment]
@cwarnermm found a couple of typos during the demo.  I have updated the two she mentioned and will wait for her to complete her review of the `en.json` file. 

Are we good to merge after she approves?  

@elias, your changes look amazin!

**Comment by jfrerich** [comment]
@enahum, your changes look great!

File: assets/base/i18n/en.json
Comment (cwarnermm) [inline assets/base/i18n/en.json]:
```suggestion
  "mobile.login_options.none": "You can't log in to your account yet. At least one login option must be configured. Contact your System Admin for assistance.",
```

File: assets/base/i18n/en.json
Comment (cwarnermm) [inline assets/base/i18n/en.json:201]:
```suggestion
  "mobile.oauth.failed_to_open_link_no_browser": "The link failed to open. Please verify that a browser is installed on the device.",
```

File: assets/base/i18n/en.json
Comment (cwarnermm) [inline assets/base/i18n/en.json:185]:
```suggestion
  "mobile.managed.jailbreak": "Jailbroken devices are not trusted by {vendor}. Login can't continue. Please exit the app.",
```

File: assets/base/i18n/en.json
Comment (cwarnermm) [inline assets/base/i18n/en.json:281]:
```suggestion
  "permalink.show_dialog_warn.description": "You are about to join {channel} without explicitly being added by the Channel Admin. Are you sure you wish to join this private channel?",
```

File: assets/base/i18n/en.json
Comment (cwarnermm) [inline assets/base/i18n/en.json:203]:
```suggestion
  "mobile.oauth.something_wrong.okButon": "OK",
```

**Review by cwarnermm — COMMENTED** [review]
Thanks! I've provided feedback inline within the en.json file. Please note that I didn't thoroughly cross-check all string suggestions against other code files in this PR for consistency.

File: assets/base/i18n/en.json
Comment (jfrerich) [reply assets/base/i18n/en.json]: Done.

File: assets/base/i18n/en.json
Comment (jfrerich) [reply assets/base/i18n/en.json:281]: Done.

File: assets/base/i18n/en.json
Comment (jfrerich) [reply assets/base/i18n/en.json]: Done.

File: assets/base/i18n/en.json
Comment (jfrerich) [reply assets/base/i18n/en.json]: done.

File: assets/base/i18n/en.json
Comment (jfrerich) [reply assets/base/i18n/en.json:185]: @elias, Carrie submitted some grammar changes that are not related to this PR, but I did make them. See her other comments.

Regarding this suggestion, I'm not sure the we need the additional suggestion.  The i18n code is loaded in `apps/init/managed_app.ts`.  My assumption is tha that the app user wouldn't feel like they cannot login, but that the app can't load or initialize.  I'd suggestion not addint this suggestion. Your thoughts?

Comment (enahum) [reply assets/base/i18n/en.json:185]: yes, this is **NOT** login related, thank you!

Comment (enahum) [reply assets/base/i18n/en.json:185]: Also, @cwarnermm can we keep grammar and stuff reviews related to the PR, you can always submit a larger PR only for the grammar stuff in a separate one.

Also one thing that can massively help is to review the actual designs as we basically copy/paste the text that appears there.

Comment (cwarnermm) [reply assets/base/i18n/en.json:185]: Absolutely! That additional wasn't intentional; I had mistakenly believed it to be part of the login updates. @jfrerich please feel free to dismiss any suggestions that aren't directly related to this login PR effort.

I welcome the opportunity to get involved earlier in the process by reviewing the designs prior to implementation. Please feel free to loop me in on the design phase of future efforts!

Comment (jfrerich) [reply assets/base/i18n/en.json:185]: If you are interested in the designs, they are located [here](https://www.figma.com/file/a9w2q3JnVvBwo2IScP8j3K/%E2%9C%85-Phase-1-Design?node-id=1544%3A91092).  I will make sure to ping you early in the code implementation process for grammar help :)  You might also reach out in the Mattermost `Feature: Gekidou & Mobile v2` channel to mention your desire to review the text of our new design!

**Pull request merged**
